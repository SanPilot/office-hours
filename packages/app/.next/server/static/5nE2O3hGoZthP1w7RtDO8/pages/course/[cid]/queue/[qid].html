<!DOCTYPE html><html><head><meta name="viewport" content="width=device-width"/><meta charSet="utf-8"/><title> Queue | Khoury Office Hours</title><meta name="next-head-count" content="3"/><link rel="preload" href="/_next/static/css/0b10e9f32b7d87c5b5ed.css" as="style"/><link rel="stylesheet" href="/_next/static/css/0b10e9f32b7d87c5b5ed.css"/><link rel="preload" href="/_next/static/css/20de8917c44a516a459d.css" as="style"/><link rel="stylesheet" href="/_next/static/css/20de8917c44a516a459d.css"/><link rel="preload" href="/_next/static/5nE2O3hGoZthP1w7RtDO8/pages/_app.js" as="script"/><link rel="preload" href="/_next/static/5nE2O3hGoZthP1w7RtDO8/pages/course/%5Bcid%5D/queue/%5Bqid%5D.js" as="script"/><link rel="preload" href="/_next/static/runtime/webpack-1c5199ff66550d26e499.js" as="script"/><link rel="preload" href="/_next/static/chunks/framework.ecbd939e3f22c21530d6.js" as="script"/><link rel="preload" href="/_next/static/chunks/fc8d00fd.158594bfc26fc793d03b.js" as="script"/><link rel="preload" href="/_next/static/chunks/commons.f581a4c8661378bd241c.js" as="script"/><link rel="preload" href="/_next/static/chunks/93f45fe113450c1fe1631ef2c1802f5d3222e556.9708f84493aff45b242d.js" as="script"/><link rel="preload" href="/_next/static/chunks/4dc19552c8c1faa5ed83aafb8fd7aab1c081565b.1236030ff06720cafd5b.js" as="script"/><link rel="preload" href="/_next/static/chunks/0849904e8c250d9d2520fa136298f4be9c59f048.96e54de04c15d71ed2ef.js" as="script"/><link rel="preload" href="/_next/static/chunks/85b0cd232aeff9ddfd69927d79f7e0a6a6909f54.48a5ac24e0f744c272d3.js" as="script"/><link rel="preload" href="/_next/static/runtime/main-e3d3bfc2aa5306ba04a6.js" as="script"/><link rel="preload" href="/_next/static/chunks/a29ae703.16fee017b3d899b4d271.js" as="script"/><link rel="preload" href="/_next/static/chunks/29107295.8007f4499acc1bea4d2b.js" as="script"/><link rel="preload" href="/_next/static/chunks/b7f72fcd4cd2872f3642182ea41968631b8348f8.76f3f8b17aac8dc8335f.js" as="script"/><link rel="preload" href="/_next/static/chunks/57fee3afb3f1bb7ad1594d628173a6c1bde98abe.1081c1c8525597342cf2.js" as="script"/><link rel="preload" href="/_next/static/chunks/85fd2704d968f5c647ae424c4371f46543b32c0c.475aa6618c25f93d0df1.js" as="script"/><link rel="preload" href="/_next/static/chunks/10e89f805a533dd145ad7cee8a4a135edce75fab.b305084e226b9b2f4c71.js" as="script"/><link rel="preload" href="/_next/static/chunks/f975bc750d17b861c287c66674152829d6196eab.8593ab6f7a8a8d1d042c.js" as="script"/><link rel="preload" href="/_next/static/chunks/65725faa2ccb61e74f472549e9cdaee947affb16.372e1928c575f3c94e68.js" as="script"/><link rel="preload" href="/_next/static/chunks/e8492d42d82540a1606962d4191afdb0a9fd5adc.539c1af7b5198ae65a7c.js" as="script"/><link rel="preload" href="/_next/static/chunks/bd1f6211abca4a80ad42608c4dead073b859a7ff.7e14e97ad48a91a1b140.js" as="script"/><link rel="preload" href="/_next/static/chunks/c55c59f0b8eea51b359ffdb8c008e8ab22d6f70e.748b44e373c982be339a.js" as="script"/><link rel="preload" href="/_next/static/chunks/4a146d395d43da622638856a99365ab139a15663.e045abf429bc88180126.js" as="script"/><link rel="preload" href="/_next/static/chunks/e8d6967f9f176b522d7946638c2c250156e8ef8e.67ea4ec0b3cc9c22d67d.js" as="script"/><style data-styled="" data-styled-version="5.1.1">.jBJAjb{width:100%;background:#ebebeb;-webkit-flex-shrink:0;-ms-flex-negative:0;flex-shrink:0;padding:12px 64px;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-pack:justify;-webkit-justify-content:space-between;-ms-flex-pack:justify;justify-content:space-between;}/*!sc*/
data-styled.g1[id="Footer__FullWidth-rge8n6-0"]{content:"jBJAjb,"}/*!sc*/
.hLZna{min-height:100vh;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;}/*!sc*/
data-styled.g2[id="_app__Layout-sc-1cfs7og-0"]{content:"hLZna,"}/*!sc*/
.lbjTcU{-webkit-flex:1 0 auto;-ms-flex:1 0 auto;flex:1 0 auto;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;}/*!sc*/
data-styled.g3[id="_app__Content-sc-1cfs7og-1"]{content:"lbjTcU,"}/*!sc*/
.gcWbDU{-webkit-flex:1;-ms-flex:1;flex:1;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;margin:0 auto;width:100%;padding:0 16px;}/*!sc*/
@media (min-width:750px){.gcWbDU{padding:0 20px;}}/*!sc*/
@media (min-width:1000px){.gcWbDU{padding:0 32px;}}/*!sc*/
@media (min-width:1500px){.gcWbDU{max-width:1500px;}}/*!sc*/
data-styled.g11[id="PageContainer__StandardPageContainer-pwm33f-0"]{content:"gcWbDU,"}/*!sc*/
.cpBoPK{border-bottom:none;}/*!sc*/
data-styled.g12[id="NavBarTabs__HorizontalMenu-fmvdkp-0"]{content:"cpBoPK,"}/*!sc*/
@media (min-width:650px){.bncCm{padding:10px 50px;font-size:16px;color:#262626;}}/*!sc*/
data-styled.g13[id="NavBarTabs__MenuItem-fmvdkp-1"]{content:"bncCm,"}/*!sc*/
.cDEDJK .ant-popover-inner-content{padding:0px;}/*!sc*/
data-styled.g14[id="ProfileDrawer__NoPaddingPopover-sc-4orw83-0"]{content:"cDEDJK,"}/*!sc*/
.bbyojt{cursor:pointer;}/*!sc*/
data-styled.g15[id="ProfileDrawer__AvatarButton-sc-4orw83-1"]{content:"bbyojt,"}/*!sc*/
.hUHqtC{padding:0px 0px;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;height:67px;z-index:1;}/*!sc*/
data-styled.g16[id="NavBar__Nav-sc-1jr3ir-0"]{content:"hUHqtC,"}/*!sc*/
.bvIZvi{position:absolute;top:0;left:0;right:0;height:67px;background:#fff;border-bottom:solid 1px #e8e8e8;}/*!sc*/
data-styled.g17[id="NavBar__NavBG-sc-1jr3ir-1"]{content:"bvIZvi,"}/*!sc*/
.dAZGwx{z-index:1;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;margin-right:20px;}/*!sc*/
data-styled.g18[id="NavBar__LogoContainer-sc-1jr3ir-2"]{content:"dAZGwx,"}/*!sc*/
.PeAxD{font-size:20px;font-weight:500;color:#262626;text-transform:capitalize;}/*!sc*/
data-styled.g19[id="NavBar__Logo-sc-1jr3ir-3"]{content:"PeAxD,"}/*!sc*/
.hanrUS{-webkit-box-flex:1;-webkit-flex-grow:1;-ms-flex-positive:1;flex-grow:1;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;-webkit-box-pack:justify;-webkit-justify-content:space-between;-ms-flex-pack:justify;justify-content:space-between;padding-left:0px;}/*!sc*/
data-styled.g20[id="NavBar__MenuCon-sc-1jr3ir-4"]{content:"hanrUS,"}/*!sc*/
@media (max-width:650px){.kSYKAK{display:none;}}/*!sc*/
data-styled.g21[id="NavBar__LeftMenu-sc-1jr3ir-5"]{content:"kSYKAK,"}/*!sc*/
@media (max-width:650px){.gAbYee{display:none;}}/*!sc*/
data-styled.g22[id="NavBar__RightMenu-sc-1jr3ir-6"]{content:"gAbYee,"}/*!sc*/
.daChbZ{height:32px;padding:6px;margin-top:8px;display:none;background:none;}/*!sc*/
@media (max-width:650px){.daChbZ{display:inline-block;}}/*!sc*/
data-styled.g23[id="NavBar__BarsMenu-sc-1jr3ir-7"]{content:"daChbZ,"}/*!sc*/
.gzuZkL{display:block;width:20px;height:2px;background:#1890ff;position:relative;}/*!sc*/
.gzuZkL:after,.gzuZkL:before{content:attr(x);width:20px;position:absolute;top:-6px;left:0;height:2px;background:#1890ff;}/*!sc*/
.gzuZkL:after{top:auto;bottom:-6px;}/*!sc*/
data-styled.g24[id="NavBar__BarsButton-sc-1jr3ir-8"]{content:"gzuZkL,"}/*!sc*/
.fQzTww{-webkit-flex:1;-ms-flex:1;flex:1;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;}/*!sc*/
data-styled.g115[id="qid__Container-xy2cp0-0"]{content:"fQzTww,"}/*!sc*/
</style></head><body><div id="__next"><div class="_app__Layout-sc-1cfs7og-0 hLZna"><div class="_app__Content-sc-1cfs7og-1 lbjTcU"><div class="PageContainer__StandardPageContainer-pwm33f-0 gcWbDU"><div class="qid__Container-xy2cp0-0 fQzTww"><nav class="NavBar__NavBG-sc-1jr3ir-1 bvIZvi"></nav><nav class="NavBar__Nav-sc-1jr3ir-0 hUHqtC"><div class="NavBar__LogoContainer-sc-1jr3ir-2 dAZGwx"><div class="NavBar__Logo-sc-1jr3ir-3 PeAxD"><span></span></div></div><div class="NavBar__MenuCon-sc-1jr3ir-4 hanrUS"><div class="NavBar__LeftMenu-sc-1jr3ir-5 kSYKAK"><ul class="ant-menu NavBarTabs__HorizontalMenu-fmvdkp-0 cpBoPK ant-menu-light ant-menu-root ant-menu-horizontal" role="menu"><li class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu" style="display:none" role="menuitem"><div class="ant-menu-submenu-title" role="button" aria-expanded="false" aria-haspopup="true"><span>···</span><i class="ant-menu-submenu-arrow"></i></div></li><li class="ant-menu-item NavBarTabs__MenuItem-fmvdkp-1 bncCm ant-menu-item-only-child" role="menuitem"><a href="/course/undefined/today">Today</a></li><li class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu" style="display:none" role="menuitem"><div class="ant-menu-submenu-title" role="button" aria-expanded="false" aria-haspopup="true"><span>···</span><i class="ant-menu-submenu-arrow"></i></div></li><li class="ant-menu-item NavBarTabs__MenuItem-fmvdkp-1 bncCm ant-menu-item-only-child" role="menuitem"><a href="/course/undefined/schedule">Schedule</a></li><li class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu" style="visibility:hidden;position:absolute" role="menuitem"><div class="ant-menu-submenu-title" role="button" aria-expanded="false" aria-haspopup="true"><span>···</span><i class="ant-menu-submenu-arrow"></i></div></li></ul></div><div class="NavBar__RightMenu-sc-1jr3ir-6 gAbYee"><div class="ProfileDrawer__AvatarButton-sc-4orw83-1 bbyojt"><span style="width:40px;height:40px;line-height:40px;font-size:17.142857142857142px;background-color:#1abc9c" class="ant-avatar ant-avatar-circle"><span class="ant-avatar-string" style="opacity:0"></span></span></div></div></div><button type="button" class="ant-btn NavBar__BarsMenu-sc-1jr3ir-7 daChbZ ant-btn-primary"><span class="NavBar__BarsButton-sc-1jr3ir-8 gzuZkL"></span></button></nav><div></div></div></div></div><footer class="Footer__FullWidth-rge8n6-0 jBJAjb"><div>Made by students @<!-- --> <a>Sandbox</a>. Source on<!-- --> <a>GitHub.</a></div><div><a>Give us feedback</a><span> or </span><a>File a bug report</a></div></footer></div></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{}},"page":"/course/[cid]/queue/[qid]","query":{},"buildId":"5nE2O3hGoZthP1w7RtDO8","nextExport":true,"autoExport":true,"isFallback":false}</script><script nomodule="" src="/_next/static/runtime/polyfills-129b8641b84b21697221.js"></script><script async="" data-next-page="/_app" src="/_next/static/5nE2O3hGoZthP1w7RtDO8/pages/_app.js"></script><script async="" data-next-page="/course/[cid]/queue/[qid]" src="/_next/static/5nE2O3hGoZthP1w7RtDO8/pages/course/%5Bcid%5D/queue/%5Bqid%5D.js"></script><script src="/_next/static/runtime/webpack-1c5199ff66550d26e499.js" async=""></script><script src="/_next/static/chunks/framework.ecbd939e3f22c21530d6.js" async=""></script><script src="/_next/static/chunks/fc8d00fd.158594bfc26fc793d03b.js" async=""></script><script src="/_next/static/chunks/commons.f581a4c8661378bd241c.js" async=""></script><script src="/_next/static/chunks/93f45fe113450c1fe1631ef2c1802f5d3222e556.9708f84493aff45b242d.js" async=""></script><script src="/_next/static/chunks/4dc19552c8c1faa5ed83aafb8fd7aab1c081565b.1236030ff06720cafd5b.js" async=""></script><script src="/_next/static/chunks/0849904e8c250d9d2520fa136298f4be9c59f048.96e54de04c15d71ed2ef.js" async=""></script><script src="/_next/static/chunks/85b0cd232aeff9ddfd69927d79f7e0a6a6909f54.48a5ac24e0f744c272d3.js" async=""></script><script src="/_next/static/runtime/main-e3d3bfc2aa5306ba04a6.js" async=""></script><script src="/_next/static/chunks/a29ae703.16fee017b3d899b4d271.js" async=""></script><script src="/_next/static/chunks/29107295.8007f4499acc1bea4d2b.js" async=""></script><script src="/_next/static/chunks/b7f72fcd4cd2872f3642182ea41968631b8348f8.76f3f8b17aac8dc8335f.js" async=""></script><script src="/_next/static/chunks/57fee3afb3f1bb7ad1594d628173a6c1bde98abe.1081c1c8525597342cf2.js" async=""></script><script src="/_next/static/chunks/85fd2704d968f5c647ae424c4371f46543b32c0c.475aa6618c25f93d0df1.js" async=""></script><script src="/_next/static/chunks/10e89f805a533dd145ad7cee8a4a135edce75fab.b305084e226b9b2f4c71.js" async=""></script><script src="/_next/static/chunks/f975bc750d17b861c287c66674152829d6196eab.8593ab6f7a8a8d1d042c.js" async=""></script><script src="/_next/static/chunks/65725faa2ccb61e74f472549e9cdaee947affb16.372e1928c575f3c94e68.js" async=""></script><script src="/_next/static/chunks/e8492d42d82540a1606962d4191afdb0a9fd5adc.539c1af7b5198ae65a7c.js" async=""></script><script src="/_next/static/chunks/bd1f6211abca4a80ad42608c4dead073b859a7ff.7e14e97ad48a91a1b140.js" async=""></script><script src="/_next/static/chunks/c55c59f0b8eea51b359ffdb8c008e8ab22d6f70e.748b44e373c982be339a.js" async=""></script><script src="/_next/static/chunks/4a146d395d43da622638856a99365ab139a15663.e045abf429bc88180126.js" async=""></script><script src="/_next/static/chunks/e8d6967f9f176b522d7946638c2c250156e8ef8e.67ea4ec0b3cc9c22d67d.js" async=""></script><script src="/_next/static/5nE2O3hGoZthP1w7RtDO8/_buildManifest.js" async=""></script><script src="/_next/static/5nE2O3hGoZthP1w7RtDO8/_ssgManifest.js" async=""></script></body></html>