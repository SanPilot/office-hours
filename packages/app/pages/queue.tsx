import styled from "styled-components";
import {
  Role,
  Question,
  OpenQuestionStatus,
  ClosedQuestionStatus,
  QuestionType,
  QuestionStatus,
  UserCourse,
} from "@template/common";
import QuestionForm from "../components/Queue/QuestionForm";
import QueueList from "../components/Queue/QueueList";
import StudentPopupCard from "../components/Queue/StudentPopupCard";
import {
  useCallback,
  useState,
  useContext,
  useEffect,
  useMemo,
  Fragment,
} from "react";
import useSWR from "swr";
import { API } from "@template/api-client";
import { ProfileContext } from "../contexts/ProfileContextProvider";

// TODO: replace this with profile role from endpoint
const ROLE: Role = Role.TA;

const Container = styled.div`
  margin: 32px 64px;
  @media (max-width: 768px) {
    margin: 32px 24px;
  }
`;

interface QueueProps {}

export default function Queue({}: QueueProps) {
  const { profile } = useContext(ProfileContext);
  const [course, setCourse] = useState<UserCourse>(null);
  const [queueId, setQueueId] = useState<number>(null);
  const [questions, setQuestions] = useState<Question[]>([]);

  // Student queue state variables
  const [isJoining, setIsJoining] = useState<boolean>(false);
  const [questionDraftId, setQuestionDraftId] = useState<number>(null);

  // TA queue state variables
  const [openPopup, setOpenPopup] = useState<boolean>(false);
  const [helpingQuestions, setHelpingQuestions] = useState<Question[]>([]);
  const [groupQuestions, setGroupQuestions] = useState<Question[]>([]);
  const [currentQuestion, setCurrentQuestion] = useState<Question>(null);

  useEffect(() => {
    if (profile) {
      const selectedCourse: UserCourse = profile.courses[0];
      setCourse(selectedCourse);
      setQueueId(selectedCourse.course.id);
    }
  }, [profile]);

  useEffect(() => {
    if (queueId) {
      getQuestions();
    }
  }, [queueId]);

  /**
   * Gets the questions for this course
   */
  const getQuestions = () => {
    if (queueId) {
      API.questions.index(queueId).then((q: Question[]) => {
        if (q) {
          setQuestions(q);

          let helping: Question[] = [];
          let group: Question[] = [];
          for (let question of q) {
            if (
              question.status === OpenQuestionStatus.Helping
              // question.taHelped &&
              // question.taHelped.id === profile.id
            ) {
              helping.push(question);
            } else {
              group.push(question);
            }
          }
          setHelpingQuestions(helping);
          setGroupQuestions(group);
        }
      });
    }
  };

  const onOpenClick = useCallback((question: Question): void => {
    setCurrentQuestion(question);
    setOpenPopup(true);
  }, []);

  const onCloseClick = useCallback((): void => {
    setOpenPopup(false);
  }, []);

  /**
   * Student functions to support queue operations.
   */

  /**
   * Creates a new Question draft for a student who has joined the queue.
   */
  const joinQueue = () => {
    setIsJoining(true);

    // API call to join queue, question marked as draft
    API.questions
      .create(queueId, {
        text: "",
        questionType: null, // endpoint needs to be changed to allow empty questionType for drafts
      })
      .then((q: Question) => {
        if (q) {
          // fetch updated question list
          getQuestions();

          setQuestionDraftId(q.id);
        }
      });
  };

  /**
   * Deletes existing Question draft for a student who has left the queue.
   */
  const leaveQueue = () => {
    setIsJoining(false);

    API.questions
      .update(queueId, questionDraftId, {
        status: ClosedQuestionStatus.Deleted,
      })
      .then((q) => {
        // fetch updated question list
        getQuestions();

        setQuestionDraftId(null);
      });
  };

  /**
   * Finishes creating a given question by updating the draft.
   */
  const finishQuestion = (text: string, questionType: QuestionType) => {
    API.questions
      .update(queueId, questionDraftId, {
        text: text,
        questionType: questionType,
        status: OpenQuestionStatus.Queued,
      })
      .then((q) => {
        if (q) {
          // fetch updated question list
          getQuestions();
          setIsJoining(false);
        }
      });
  };

  /**
   * TA functions to support queue operations
   */

  /**
   * Updates a given question to the given status.
   * @param question the question being modified
   * @param status the updated status
   */
  const updateQuestionTA = (question: Question, status: QuestionStatus) => {
    API.questions
      .update(queueId, question.id, {
        status: status,
      })
      .then((q) => {
        if (q) {
          // fetch updated question list
          getQuestions();
          // fetch updated helping list
          setCurrentQuestion(q);
          // update helping state if none left
        }
      });
  };

  /**
   * Sends a push notification to the student with the given Question
   * @param question the question to be notified
   */
  const alertStudent = (question: Question) => {
    // Send API request to trigger notification
  };

  return (
    <Container>
      {!isJoining && (
        <Fragment>
          <QueueList
            role={ROLE}
            onOpenClick={onOpenClick}
            joinQueue={joinQueue}
            updateQuestionTA={updateQuestionTA}
            alertStudent={alertStudent}
            questions={questions}
            helpingQuestions={helpingQuestions}
            groupQuestions={groupQuestions}
          />
          {ROLE === "ta" && currentQuestion && (
            <StudentPopupCard
              onClose={onCloseClick}
              email="takayama.a@northeastern.edu" //need a way to access this. or the user
              wait={20} //figure out later
              question={currentQuestion}
              location="Outside by the printer" // need a way to access this
              visible={openPopup}
              updateQuestion={updateQuestionTA}
            />
          )}
        </Fragment>
      )}
      {isJoining && (
        <QuestionForm leaveQueue={leaveQueue} finishQuestion={finishQuestion} />
      )}
    </Container>
  );
}
