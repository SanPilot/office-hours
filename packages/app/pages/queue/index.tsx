import styled from "styled-components";
import QueueList from "../../components/Queue/QueueList";

const Container = styled.div`
  margin: 32px 64px;
  @media (max-width: 768px) {
    margin: 32px 24px;
  }
`;

interface QueueProps {}

export default function Queue({}: QueueProps) {
  return (
    <Container>
      <QueueList />
    </Container>
  );
}
