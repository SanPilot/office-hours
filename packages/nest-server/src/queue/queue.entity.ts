import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  ManyToOne,
  JoinColumn,
  OneToMany,
} from "typeorm";
import { Course } from "../course/course.entity";
import { Question } from "../question/question.entity";

@Entity("queue_model")
export class Queue extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne((type) => Course, (course) => course.queues)
  @JoinColumn({ name: "courseId" })
  course: Course;

  @Column({ nullable: true })
  courseId: number;

  @Column("text")
  room: string;

  @OneToMany((type) => Question, (qm) => qm.queue)
  questions: Question[];

  // TODO: eventually figure out how staff get sent to FE as well
}
