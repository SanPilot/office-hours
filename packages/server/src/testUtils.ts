import { createConnection, Connection } from "../../ormconfig";
import { init } from "./server";
import path from "path";
import { Server } from "@hapi/hapi";

export function setupServerTest(): () => Server {
  let server: Server;

  beforeEach(async () => {
    server = await init();
  });

  afterEach(async () => {
    await server.stop();
  });

  return () => server;
}

export function setupDBTest() {
  let connection: Connection;
  beforeAll(async () => {
    connection = await createConnection({
      type: "postgres",
      url: "postgres://postgres@localhost:5432/test",
      synchronize: true,
      logging: false,
      entities: [path.resolve(__dirname, "entity/**/*{.js,.ts}")],
      migrations: [path.resolve(__dirname, "migration/**/*{.js,.ts}")],
      subscribers: [path.resolve(__dirname, "subscriber/**/*{.js,.ts}")],
    });
  });

  beforeEach(async () => {
    await connection.synchronize(true);
  });

  afterAll(async () => {
    await connection.close();
  });
}
