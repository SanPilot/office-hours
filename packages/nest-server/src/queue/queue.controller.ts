import { Controller, Get, Param, NotFoundException } from '@nestjs/common';
import { Connection } from 'typeorm';
import { Queue } from './queue.entity';
import {
  GetQueueResponse,
  UserPartial,
  User,
  QuestionStatus,
  QuestionStatusKeys,
  Question as QuestionDTO,
  ListQuestionsResponse,
} from '@template/common';
import { Question } from '../question/question.entity';

@Controller('queues')
export class QueueController {
  constructor(private connection: Connection) {}
  @Get(':queueId')
  async getQueue(@Param('queueId') queueId: string): Promise<GetQueueResponse> {
    const queue = await Queue.findOne(queueId);

    const queueSize = await Question.count({
      where: { queueId: queue.id },
    });

    // TODO: Fill this in with real data
    const staffList = [
      {
        id: 54,
        name: 'Will Stenzel',
        photoURL:
          'https://prod-web.neu.edu/wasapp/EnterprisePhotoService/PhotoServlet?vid=CCS&er=24f4b12cccbf875c7740bbfed45a993900cf0d08d11aa07c84780b3a3501f3bacca4eb33ed5effee8aa2dd195750cfbc9884dd5f2ac62c8f',
      },
      {
        id: 42,
        name: 'Grisha Zaytsev',
        photoURL:
          'https://prod-web.neu.edu/wasapp/EnterprisePhotoService/PhotoServlet?vid=CCS&er=471f2d695fbb8a00ee740ad3ea910453986aec81ddaecf889ae98b3a1858597b12650afd0d4e59c561172f76cb1946eec217ed89bd4074c0',
      },
    ];

    return {
      id: queue.id,
      room: queue.room,
      staffList: staffList,
      queueSize: queueSize,
    };
  }

  @Get(':queueId/questions')
  async getQuestions(
    @Param('queueId') queueId: string,
  ): Promise<ListQuestionsResponse> {
    // todo: need a way to return different data, if TA vs. student hits endpoint.
    // for now, just return the student response
    const queueSize = await Queue.count({
      where: { id: queueId },
    });
    // Check that the queue exists
    if (queueSize === 0) {
      throw new NotFoundException();
    }

    const questions = await Question.find({
      where: [
        {
          queueId: queueId,
        },
      ],
      relations: ['creator', 'taHelped'],
    });

    return questions.map(questionModelToQuestion);
  }
}

function questionModelToQuestion(qm: Question): QuestionDTO {
  return {
    creator: userModelToUserPartial(qm.creator),
    id: qm.id,
    createdAt: qm.createdAt,
    status: parseStatus(qm.status),
    text: qm.text,
    // qm.taHelped: types says is nonnullable, but it is nullable
    taHelped: qm.taHelped ? userModelToUserPartial(qm.taHelped) : null,
    closedAt: qm.closedAt,
    questionType: qm.questionType,
    // TODO: helpedAt: property not required in types, but required by JOI
    helpedAt: qm.helpedAt,
  };
}

function userModelToUserPartial(um: User): UserPartial {
  return {
    id: um.id,
    name: um.name,
    // TODO: photoURL: property not required in types, but required by JOI
    photoURL: um.photoURL,
  };
}

function parseStatus(maybeStatus: string | number | symbol): QuestionStatus {
  if (maybeStatus in QuestionStatusKeys) {
    return maybeStatus as QuestionStatus;
  } else {
    throw new Error(
      `received unknown or ill-formatted status: ${String(maybeStatus)}`,
    );
  }
}
