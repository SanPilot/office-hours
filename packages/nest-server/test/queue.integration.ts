import { QueueFactory, QuestionFactory } from './util/factories';
import { setupIntegrationTest } from './util/testUtils';
import { QueueModule } from '../src/queue/queue.module';

describe('Queue Integration', () => {
  const supertest = setupIntegrationTest(QueueModule);

  describe('GET /queues/:id', () => {
    it('get a queue', async () => {
      const queue = await QueueFactory.create({
        questions: [await QuestionFactory.create()],
      });
      const res = await supertest().get(`/queues/${queue.id}`).expect(200);
      expect(res.body).toMatchSnapshot();
    });

    it('returns 401 when not logged in', () => {});
  });

  describe('GET /queues/:id/questions', () => {
    it('returns questions in the queue', async () => {
      const queue = await QueueFactory.create({
        questions: [await QuestionFactory.create({ text: 'in queue' })],
      });
      await QuestionFactory.create({ text: 'not in queue' });

      const res = await supertest()
        .get(`/queues/${queue.id}/questions`)
        .expect(200);
      expect(res.body).toMatchSnapshot();
    });
  });
});
