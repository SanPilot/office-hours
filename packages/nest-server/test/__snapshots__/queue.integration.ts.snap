// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queue Integration GET /queues/:id get a queue 1`] = `
Object {
  "id": 2,
  "queueSize": 1,
  "room": "WVH 2",
  "staffList": Array [
    Object {
      "id": 54,
      "name": "Will Stenzel",
      "photoURL": "https://prod-web.neu.edu/wasapp/EnterprisePhotoService/PhotoServlet?vid=CCS&er=24f4b12cccbf875c7740bbfed45a993900cf0d08d11aa07c84780b3a3501f3bacca4eb33ed5effee8aa2dd195750cfbc9884dd5f2ac62c8f",
    },
    Object {
      "id": 42,
      "name": "Grisha Zaytsev",
      "photoURL": "https://prod-web.neu.edu/wasapp/EnterprisePhotoService/PhotoServlet?vid=CCS&er=471f2d695fbb8a00ee740ad3ea910453986aec81ddaecf889ae98b3a1858597b12650afd0d4e59c561172f76cb1946eec217ed89bd4074c0",
    },
  ],
}
`;

exports[`Queue Integration GET /queues/:id/questions returns questions in the queue 1`] = `
Array [
  Object {
    "closedAt": null,
    "createdAt": "2020-07-05T01:55:48.706Z",
    "creator": Object {
      "id": 1,
      "name": "John Doe the 2th",
      "photoURL": "https://pics/2",
    },
    "helpedAt": null,
    "id": 1,
    "questionType": "Other",
    "status": "Queued",
    "taHelped": null,
    "text": "in queue",
  },
]
`;
